# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: netwerk/protocol/http/nsHttpTransaction.cpp
# Commit: b3ffe3409712
# Full Hash: b3ffe3409712f31bcf61cf93aa09bf137e45f9bc
# Author: Kershaw Chang <kershaw@mozilla.com>
# Date: 2025-03-29 09:19:43
# Description:
#   Bug 1951554 - Avoid racing on nsHttpTransaction::mConnInfo, r=necko-reviewers,valentin
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D243146
# ==============================================================================

diff -r 21cdd98a6828 -r b3ffe3409712 netwerk/protocol/http/nsHttpTransaction.cpp
--- a/netwerk/protocol/http/nsHttpTransaction.cpp	Fri Mar 28 17:48:02 2025 +0000
+++ b/netwerk/protocol/http/nsHttpTransaction.cpp	Fri Mar 28 17:56:38 2025 +0000
@@ -217,6 +217,7 @@
   if (NS_FAILED(rv)) return rv;
 
   mConnInfo = cinfo;
+  mFinalizedConnInfo = cinfo;
   mCallbacks = callbacks;
   mConsumerTarget = target;
   mCaps = caps;
@@ -438,12 +439,18 @@
   }
 }
 
-UniquePtr<nsHttpResponseHead> nsHttpTransaction::TakeResponseHead() {
+UniquePtr<nsHttpResponseHead> nsHttpTransaction::TakeResponseHeadAndConnInfo(
+    nsHttpConnectionInfo** aOut) {
   MOZ_ASSERT(!mResponseHeadTaken, "TakeResponseHead called 2x");
 
   // Lock TakeResponseHead() against main thread
   MutexAutoLock lock(mLock);
 
+  if (aOut) {
+    RefPtr<nsHttpConnectionInfo> connInfo = mFinalizedConnInfo;
+    connInfo.forget(aOut);
+  }
+
   mResponseHeadTaken = true;
 
   // Even in OnStartRequest() the headers won't be available if we were
@@ -523,6 +530,7 @@
 
   mActivated = true;
   gHttpHandler->ConnMgr()->AddActiveTransaction(this);
+  FinalizeConnInfo();
 }
 
 void nsHttpTransaction::GetSecurityCallbacks(nsIInterfaceRequestor** cb) {
@@ -1830,6 +1838,14 @@
   }
 }
 
+void nsHttpTransaction::FinalizeConnInfo() {
+  RefPtr<nsHttpConnectionInfo> cloned = mConnInfo->Clone();
+  {
+    MutexAutoLock lock(mLock);
+    mFinalizedConnInfo.swap(cloned);
+  }
+}
+
 void nsHttpTransaction::SetRestartReason(TRANSACTION_RESTART_REASON aReason) {
   if (mRestartReason == TRANSACTION_RESTART_NONE) {
     mRestartReason = aReason;